File: src/cs/PhantasyIslandPythonRemoteControl/radio/WaitToken.cs
using System;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace PhantasyIslandPythonRemoteControl.Radio
{
    public class WaitToken
    {
        public string WaitCmd { get; }
        public long TimeBaseId { get; }
        public JsonElement? Response { get; private set; }
        public object? ProcessedResponse { get; private set; }

        private readonly ManualResetEvent _event = new ManualResetEvent(false);
        private readonly TaskCompletionSource<object?> _tcs = new TaskCompletionSource<object?>();
        private Func<JsonElement, object>? _postProcessor;

        public WaitToken(string waitCmd, long timeBaseId)
        {
            WaitCmd = waitCmd;
            TimeBaseId = timeBaseId;
        }

        public bool IsDone => _event.WaitOne(0);

        public WaitToken SetPostProcessor(Func<JsonElement, object> processor)
        {
            _postProcessor = processor;
            return this;
        }

        private object ApplyPostProcessor(JsonElement data)
        {
            if (_postProcessor != null)
            {
                ProcessedResponse = _postProcessor(data);
            }
            else
            {
                ProcessedResponse = data;
            }

            return ProcessedResponse;
        }

        public void Complete(JsonElement data)
        {
            Response = data;
            var result = ApplyPostProcessor(data);
            _event.Set();
            _tcs.TrySetResult(result);
        }

        public object? Wait(TimeSpan timeout)
        {
            if (_event.WaitOne(timeout))
            {
                return ProcessedResponse;
            }

            return null;
        }

        public Task<object?> WaitAsync() => _tcs.Task;
    }
}

File: src/fs/WaitToken.fs
namespace PhantasyIslandRemoteControl.Radio

open System
open System.Text.Json
open System.Threading
open System.Threading.Tasks

type WaitToken(waitCmd: string, timeBaseId: int64) =
    let _event = new ManualResetEvent(false)
    let _tcs = TaskCompletionSource<obj>()
    let mutable _postProcessor: Func<JsonElement, obj> = null
    
    member val WaitCmd = waitCmd with get
    member val TimeBaseId = timeBaseId with get
    member val Response: Nullable<JsonElement> = Nullable() with get, set
    member val ProcessedResponse: obj = null with get, set

    member this.IsDone = _event.WaitOne(0)

    member this.SetPostProcessor(processor: Func<JsonElement, obj>) =
        _postProcessor <- processor
        this

    member private this.ApplyPostProcessor(data: JsonElement) =
        if _postProcessor <> null then
            this.ProcessedResponse <- _postProcessor.Invoke(data)
        else
            this.ProcessedResponse <- data
        this.ProcessedResponse

    member this.Complete(data: JsonElement) =
        this.Response <- Nullable(data)
        let result = this.ApplyPostProcessor(data)
        _event.Set() |> ignore
        _tcs.TrySetResult(result) |> ignore

    member this.Wait(timeout: TimeSpan) =
        if _event.WaitOne(timeout) then
            this.ProcessedResponse
        else
            null

    member this.WaitAsync() = _tcs.Task
